Features/veen 3 loading tag types

gijskant opened the pull request:
- Exporting browse tags.
- Loading i2b2 tags.
- Conversion script from browse tag types to i2b2 tag types.

cataphract commented:
I just had a cursory look, but for now a few comments:

* The build needs to pass.
* This should go without saying, but it needs tests.
* I see database queries strewn through a lot of classes; I think this would benefit from some refactoring, having them put in a DAO.
* The (untested) Python script seems unnecessary. Why not to output the data already in the correct format? It seems the only point of exporting the browse tags is to import them as concept tags (unless we want to reimplement the export functionality of transmart here).

cataphract commented on src/main/groovy/org/transmartproject/batch/browsetag/BrowseTagAssociationToI2b2Writer.groovy:
```diff
@@ -0,0 +1,58 @@
+                    void writeHeader(Writer writer) throws IOException {
+                        writer.write(['concept_key', 'tag_title', 'tag_description', 'index'].join('\t'))
+                    }
+                }
```
In Groovy this is done with `fieldExtractor: { ['\\', it.value.type.displayName, it.value.description, 0] } as FieldExtractor` and `headerCallback: { it.write(['concept_key', 'tag_title', 'tag_description', 'index'].join('\t')) } as FlatFileHeaderCallback`.

Plus, there's no need to have a wrapper class just for instantiating a  `FlatFileItemWriter`. Either do this instantiation directly in the `@Configuration` class, or, if you really want to have a separate class, implement `FactoryBean`.

cataphract commented:
Please rebase on master, as it has a fix for a connection leak on Oracle which causes spurious failures.

gijskant commented:
- [x] Rebase on master
- [x] Add functional tests for tag type and tag import.
- [ ] Add functional tests for browse tag export.
- [x] The build needs to pass for Postgres.
- [ ] The build needs to pass for Oracle.
- [ ] Refactoring code to have database queries separate in a DAO.
- [ ] Change the browse tags export to write in the tag types import format (multiple values comma-separated on a single line)

gijskant commented:
@cataphract Comments addressed. Build passes for Postgres, not for Oracle, because there are no Oracle table definitions for tag types in transmart-data.

forus commented on src/main/groovy/org/transmartproject/batch/tag/TagTypeWriter.groovy:
```diff
@@ -0,0 +1,24 @@
+    @Override
+    void write(List<? extends TagType> items) throws Exception {
+        insertTagTypeService.deleteAllTagsForTagTypes()
+        insertTagTypeService.deleteAllTagTypes()
```
@gijskant The cleanup of the tables in this place doesn't look right. The `write` method would be executed multiple times (for each chunk) during the upload. At the end you would get only items written down at the last chunk.

gijskant replied:
Okay, I'll make it a separate step/tasklet.

cataphract commented:
BTW Coverage report is here: https://codecov.io/github/thehyve/transmart-batch/commit/851162933c82e4a64421b3838edb593efcdc41d4

It's not been posted automaticaly because of the build failure. There's a bunch of uncovered code in `BrowseTagAssociationDatabaseReader`.

cataphract commented:
It seems you only allow type `ANALYZED_STRING`. Fields of this type are not appropriate for the showing up in the filters. Only non-analyzed string. I'll write my code assuming the values `DATE`, `NON_ANALYZED_STRING`, `ANALYZED_STRING`, `INTEGER` and `FLOAT`.

cataphract commented:
Also the convention I used for the field names are that they are lowercase except for predefined ones like the catch-all `TEXT`. I'll automatically lowercase them, but it would be good if the examples used lowercase as well and not that this field is case insensitive.

cataphract commented on global/tagtypes/tagtypes.txt:
```diff
@@ -0,0 +1,3 @@
+node_type	title	solr_field_name	value_type	shown_if_empty	values	index
+STUDY	Test tag	TEST_TAG	ANALYZED_STRING	Y	Test option 1,Test option 2,Test option 3	1
+STUDY	Programming language	PROGRAMMING_LANGUAGE	ANALYZED_STRING	N	Java,C,R,Python,Javascript,Pascal,Haskell	2
```
Rename this tsv.

cataphract commented on src/test-func/groovy/org/transmartproject/batch/browsetag/BrowseTagExportJobTests.groovy:
```diff
@@ -0,0 +1,88 @@
+        directory.listFiles().each {
+            File f -> log.warn("File present: $f.absolutePath")
+        }
+    }
```
Use the temporary files `Rule`.

cataphract commented on src/main/groovy/org/transmartproject/batch/browsetag/BrowseTagTypeDatabaseReader.groovy:
```diff
@@ -0,0 +1,181 @@
+    @SuppressWarnings('UnusedPrivateMethodParameter')
+    private BrowseTagType mapRow(ResultSet rs, int rowNum) throws SQLException {
+        def tagType = getTagType(rs)
+        // FIXME: fetch values
```
What's to be fixed here?

cataphract commented on src/main/groovy/org/transmartproject/batch/browsetag/BrowseTagTypeDatabaseReader.groovy:
```diff
@@ -0,0 +1,181 @@
+        tagType
+    }
+
+}
```
Method ordering in this method seems completely random.

cataphract commented on src/main/groovy/org/transmartproject/batch/tag/DeleteTagTypesTasklet.groovy:
```diff
@@ -0,0 +1,27 @@
+        log.info("Deleting tag types...")
+        tagTypeService.deleteAllTagsForTagTypes()
+        tagTypeService.deleteAllTagTypes()
+        RepeatStatus.FINISHED
```
Should update the counts to indicate to the user how many items were deleted. Since you actually have two types of items, you can return the deleted tag types or some sort of total (logging at INFO the number for each type); or separate in two steps.

cataphract commented on src/main/groovy/org/transmartproject/batch/tag/InvalidInput.groovy:
```diff
@@ -0,0 +1,7 @@
+/**
+ * Invalid input.
+ */
+class InvalidInput extends RuntimeException {
```
Exceptions conventionally end in `Exception`. Adding this exception also seems unnecessary. See comment on `setShownIfEmpty`.

cataphract commented on src/main/groovy/org/transmartproject/batch/tag/InvalidTagDescription.groovy:
```diff
@@ -0,0 +1,12 @@
+ * Note that this is a Spring BATCH validation exception
+ */
+@InheritConstructors
+class InvalidTagDescription extends ValidationException {
```
Exceptions conventionally end in `Exception`.

cataphract commented on src/main/groovy/org/transmartproject/batch/tag/TagTypeFlatFileReader.groovy:
```diff
@@ -0,0 +1,60 @@
+/**
+ *
+ */
+class TagTypeTsvFileReader implements ItemStreamReader<TagType>, StepBuildingConfigurationTrait  {
```
Don't use `StepBuildingConfigurationTrait`... This is to be used in `@Configuration` classes, as evidenced by all the methods that it contains and that you're copying to this class. Fixing this is simple, create a factory method in a `@Configuration` class and make `TagTypeMapper` a top level (it doesn't have any dependence on the outer class here, and in fact should have been marked `static`), which is something that you seem oddly reluctant to do, judging from all the classes that are mere wrappers around the real `@Delegate`d artifact.

cataphract commented on src/main/groovy/org/transmartproject/batch/tag/TagTypeResource.groovy:
```diff
@@ -0,0 +1,73 @@
+    @Autowired
+    JdbcTemplate jdbcTemplate
+
+    class TagTypeToOptionsMapper implements RowMapper<String> {
```
Should be proivate static

cataphract commented on src/main/groovy/org/transmartproject/batch/tag/TagTypeResource.groovy:
```diff
@@ -0,0 +1,73 @@
+
+    }
+
+    private final Map<String, Map<String, Integer>> tagTypeToOptionsMap = [:]
```
Put the fields next to each other, preferably on top.

cataphract commented on src/main/groovy/org/transmartproject/batch/tag/TagTypeResource.groovy:
```diff
@@ -0,0 +1,73 @@
+
+    private final Map<String, Map<String, Integer>> tagTypeToOptionsMap = [:]
+
+    @PostConstruct
```
This should be done on a restart-if-complete step, not on the application context initialization (!).

Also, as I mentioned before, I think the database access methods that are now in `TagTypeResource` and `TagTypeService` should be put in one method (you can of course still have a bean where the list of tags options is stored)

cataphract commented on src/main/groovy/org/transmartproject/batch/tag/TagTypeService.groovy:
```diff
@@ -0,0 +1,113 @@
+    @Value(Tables.I2B2_TAG_OPTIONS)
+    SimpleJdbcInsert insertTagOptions
+
+    class TagTypeToIdMapper implements RowMapper<String> {
```
Should be `private static`.

gijskant commented:
@cataphract Now also adding/removing tag type options is supporting, without deleting existing references to unchanged options.
Tested for both Postgres and Oracle.

gijskant commented:
@forus @cataphract Could you have another look at this pull request? I think all major issues have been resolved.

cataphract commented:
This is a 2000 line changeset. Do you have a diff against the last thing I reviewed?

gijskant commented:
Here is a diff, I think it covers the changes since the last review. But that's also 1461 lines...
[diff.txt](https://github.com/thehyve/transmart-batch/files/263510/diff.txt)

cataphract commented on src/main/groovy/org/transmartproject/batch/browsetag/BrowseTagTypeDatabaseReader.groovy:
```diff
@@ -0,0 +1,180 @@
+
+    private final Map<String, BrowseFolderType> folderTypes = [:]
+
+    private final Map<Long, BrowseTagType> tagTypes = [:]
```
Make it clear that these properties store caches. If they were not classes, you'd have to save them to satisfy the contract of `ItemStreamReader`, which was my first though.

This bean should also be make step scoped, as you otherwise you're wasting memory by having these caches around.

cataphract commented on src/main/groovy/org/transmartproject/batch/browsetag/BrowseTagTypeDatabaseReader.groovy:
```diff
@@ -0,0 +1,180 @@
+        while ((value = valueReader.read()) != null) {
+            values.add(value)
+        }
+        valueReader.close()
```
The connection and the result set leak if there's an exception during the loop. I also don't get why you're manually managing spring batch resources when you could just have used the (named) jdbcTemplate.

cataphract commented on src/main/groovy/org/transmartproject/batch/tag/TagTypeWriter.groovy:
```diff
@@ -0,0 +1,54 @@
+
+        tagTypeService.insert(insertItems)
+
+        tagTypesJobMetadata.writtenTagTypes.addAll(items*.title)
```
The implementation is faulty across restarts. You add the items here to `tagTypesJobMetadata`, but `tagTypesJobMetadata` doesn't survive restarts. Which means that delete step that comes next (that just had all the **current** tags loaded from the database) will delete all the items that were written in the previous job execution.

gijskant commented:
@cataphract The Oracle tests seem to fail due to a new SSL certificate on `oracle-launcher.thehyve.net`: https://travis-ci.org/thehyve/transmart-batch/jobs/130800947#L330.

gijskant commented:
@cataphract @forus I updated `transmart-batch`. I squashed the changes, but there is a branch `/features/VEEN-3_loading_tag_types_unsquashed` with the changes before squashing: https://github.com/thehyve/transmart-batch/tree/features/VEEN-3_loading_tag_types_unsquashed
